File: 알고리즘/유찬홍/2512.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    int n, m, arr[10000], max = 0;
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
        if (max < arr[i]) max = arr[i];
    }
    scanf("%d", &m);
    int s = 0, e = max, rs = 0;
    while (s <= e) {
        int mid = (s + e) / 2, sum = 0;
        for (int i = 0; i < n; i++) {
            sum += arr[i] - mid < 0 ? arr[i] : mid;
        }
        if (sum > m) e = mid - 1;
        else {
            if (rs < mid) rs = mid;
            s = mid + 1;
        }
    }
    printf("%d", rs);
    return 0;
}

File: 알고리즘/정지관/javascript/2512.js
let fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');
let n = Number(input[0])
let arr = input[1].split(' ').map(Number).sort((a, b) => b - a)
let m = Number(input[2])

let start = 1; // 이진 탐색을 위한 시작점(start)과 끝점(end) 설정
let end = arr.reduce((a, b) => Math.max(a, b));

let result = 0;

while(start <= end) {
    let mid = parseInt((start + end) / 2)
    let total = 0;
    for(let x of arr) {
        total += Math.min(mid, x)
    }
    if(total <= m) {
        result = mid;
        start = mid + 1;
    } else {
        end = mid - 1;
    }
}


console.log(result)
